src/Api/HttpRestClient.php

```diff
@@ -0,0 +1,365 @@
+<?php
+
+declare(strict_types=1);
+
+namespace KafkaRest\Api;
+
+use Assert\Assertion;
+use Http\Discovery\Psr17FactoryDiscovery;
+use Http\Discovery\Psr18ClientDiscovery;
+use JMS\Serializer\SerializerBuilder;
+use JMS\Serializer\SerializerInterface;
+use KafkaRest\Api\Exception\UnexpectedApiResponse;
+use KafkaRest\Api\Value\Request\ConsumerOptions;
+use KafkaRest\Api\Value\Request\Subscription;
+use KafkaRest\Api\Value\Response\AssignedPartition;
+use KafkaRest\Api\Value\Response\Consumer;
+use KafkaRest\Api\Value\Response\Message;
+use KafkaRest\Api\Value\Response\Offset;
+use KafkaRest\Api\Value\Response\Partition;
+use KafkaRest\Api\Value\Response\ProduceResults;
+use KafkaRest\Api\Value\Response\Topic;
+use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Psr\Http\Message\UriFactoryInterface;
+use Psr\Http\Message\UriInterface;
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Teapot\StatusCode\Http;
+use function http_build_query;
+use function Safe\json_decode;
+use function Safe\sprintf;
+
+final class HttpRestClient implements RestClient
+{
+    /** @var LoggerInterface */
+    private $logger;
+
+    /** @var ClientInterface */
+    private $client;
+
+    /** @var UriFactoryInterface */
+    private $uriFactory;
+
+    /** @var RequestFactoryInterface */
+    private $requestFactory;
+
+    /** @var StreamFactoryInterface */
+    private $streamFactory;
+
+    /** @var SerializerInterface */
+    private $serializer;
+
+    /** @var UriInterface */
+    private $apiUri;
+
+    /** {@inheritDoc} */
+    public function __construct(
+        string $apiUri,
+        ?LoggerInterface $logger = null,
+        ?ClientInterface $client = null,
+        ?UriFactoryInterface $uriFactory = null,
+        ?RequestFactoryInterface $requestFactory = null,
+        ?StreamFactoryInterface $streamFactory = null,
+        ?SerializerInterface $serializer = null
+    ) {
+        $this->logger = $logger ?? new NullLogger();
+        $this->client = $client ?? Psr18ClientDiscovery::find();
+        $this->uriFactory = $uriFactory ?? Psr17FactoryDiscovery::findUrlFactory();
+        $this->requestFactory = $requestFactory ?? Psr17FactoryDiscovery::findRequestFactory();
+        $this->streamFactory = $streamFactory ?? Psr17FactoryDiscovery::findStreamFactory();
+        $this->serializer = $serializer ?? SerializerBuilder::create()->build();
+
+        $this->apiUri = $this->uriFactory->createUri($apiUri);
+    }
+
+    /** {@inheritDoc} */
+    public function listTopics() : array
+    {
+        /** @var array<string> $topics */
```

Comment:
is this for phpstan? maybe custom extension would clean those up

Reply:
Yes, it is. What custom extension? Like just for this library? Don't see a need for that ... I would have to put in the same amount of information + a lot of boilerplate around it. This is more transparent.

Reply:
Yup, just for this lib. Seen people doing more and more often these days. W.e